### DISCUSSION TITLE
UTF-8 URI path issues in getStaticPaths()

### DISCUSSION THREAD

**Discussion by johnny-kc-wu** [vercel/next.js#17620] [Help] [answered]
I am working on a web app showing some clips. For each clip, there's a main page including some meta data and hashtags. When a user click on the tag, it brings them to the tag page showing all clips related by using the Dynamic Routes like `pages/t/[tag].js`. The tag names may contain some CJK words (UTF-8 only) and space in between, and it worked fine until recently I have ugraded next.js from v9.4 to v9.5.3. The `getStaticPaths()` can still load the whole tags from my backend API, but this time it throw a default 404 page immediately. The following are my dependencies and the code:

```json
{
  "dependencies": {
    "@material-ui/core": "^4.11.0",
    "@material-ui/icons": "^4.9.1",
    "@material-ui/lab": "^4.0.0-alpha.56",
    "next": "9.5.3",
    "react": "16.13.1",
    "react-dom": "16.13.1",
    "recoil": "0.0.13",
    "swr": "^0.3.2",
    "uuid": "^8.3.0"
  },
}
```

```javascript
// SingleClipTags.js: the clip main page UI showing clickable hashtags

import Link from "next/link";
import { Chip } from "@material-ui/core";
...
<Link href={`/t/${tag}`} passHref>
                  <Chip
                    color="primary"
                    label={`🐶 ${tag}`}
                    className={classes.grid_chip}
                    clickable
                  />
</Link>
...
```

```javascript
// /pages/t/[tag].js: The tag page with dynamic routes

import React from "react";
import Link from "next/link";
import { useRouter } from "next/router";
import PrimarySearchAppBar from "../../src/AppBar";
import { Container, Grid, Typography } from "@material-ui/core";

export default function Tags({ anis }) {
  const router = useRouter();

  const listItems = anis.data_list.map((data) => (
    <li key={data.resource_id}>
      <Grid item>
        <Link href={`/g/${data.url_path_name}`}>
          <a>
            <picture>
              // a clip
            </picture>
          </a>
        </Link>
      </Grid>
    </li>
  ));

  return (
    <>
      <PrimarySearchAppBar />
      <Container maxWidth="md">
        <div>
          <Typography variant="h1">
            👉{router.asPath.substring(3)}👈
          </Typography>
        </div>
        <Grid
          container
          justify="space-evenly"
          component="ul"
        >
          {listItems}
        </Grid>
      </Container>
    </>
  );
}

// ### Static Generation ###
export async function getStaticPaths() {
  // url = my backend API Endpoint
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      charset: "UTF-8",
    },
  });

  if (!res.ok) {
    const error = new Error("something wrong!");
    error.info = await res.json();
    error.status = res.status;
    throw error;
  }
  const rawData = await res.json();
  /* rawData will have a tag name list like:
{
  "cursor": null,
  "data_list": [
    {
      "created_time": "2020-10-03T10:15:34.534231+00:00",
      "tag_name": "那須の長楽寺"
    },
    {
      "created_time": "2020-10-03T10:15:34.534231+00:00",
      "tag_name": "秋田犬"
    },
    ...
    ...
  ]
}
*/
  
  const paths = rawData.data_list.map((data) => ({
    params: { tag: data.tag_name },
  }));

  return { paths, fallback: false };
}

export async function getStaticProps({ params }) {
  console.log("params: ", params);
  // url = my backend API Endpoint
 /* params.tag contains data like:
  {
	"tag_name": "貓貓"
  }
*/ 
  const jsonData = {
    tag_name: params.tag,
  };
 
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      charset: "UTF-8",
    },
    body: JSON.stringify(jsonData),
  });

  if (!res.ok) {
    const error = new Error("something wrong!");
    error.info = await res.json();
    error.status = res.status;
    throw error;
  }

  const anis = await res.json();
/*
anis will have data list like:
{
  "cursor": null,
  "data_list": [
    {
      "created_time": "2020-10-03T10:15:34.534231+00:00",
      "resource_id": "an id",
      "s3_obj_key": "an obj key",
      "url_path_name": "a clip page uri"
    },
    {
      "created_time": "2020-09-05T04:08:53.123718+00:00",
      "resource_id": "an id",
      "s3_obj_key": "an obj key",
      "url_path_name": "another clip page uri"
    },
    ...
    ...
  ]
}
*/
  return { props: { anis } };
}

```
I also found that the tag names which contain plain english without any punctuation in between, for examples, words like `maru` or `cat`, work fine in my app. However, once I click on the tag with the name contain words like `my maru`, `my cat`, or CJK words like `那須の長楽寺`, `秋田犬`, the default 404 page shows up again, even though `getStaticPaths()` did receive those tag names from my backend API.

In the end, I put `encodeURIComponent()` in `getStaticPaths()` before the `paths` is going to return and it finally works!!
```javascript
//  /pages/t/[tag].js
...
const paths = rawData.data_list.map((data) => ({
    params: { tag: encodeURIComponent(data.tag_name) },
  }));
...
```
But, I still can't figure out why I need this encode function, cuz back in v9.4, the app worked fine without it. Did I miss something in v9.5? More importantly I'd like to know if I'm doing the Static Generation with dynamic routes in the right way for my situation, thanks!

**Comment by tyteen4a03** [comment]
I can trigger this bug as well.

Might be related to https://github.com/vercel/next.js/issues/10084.

Can the team please add test cases for non-ASCII URLs?

**Comment by ijjk** [comment, marked as answer]
Hi, this has been updated in the latest version of Next.js `v10.0.5-canary.8`, specifically [this PR](https://github.com/vercel/next.js/pull/19135), please update and give it a try!

**Reply by tyteen4a03** [reply]
Yep, looks like that fixed it!
